File: Software/Library/C++C2SIMClientLibv4.8.0.3/C2SIMSTOMPMessage.cpp
/*----------------------------------------------------------------*
|   Copyright 2009-2020 Networking and Simulation Laboratory      |
|         George Mason University, Fairfax, Virginia              |
|                                                                 |
| Permission to use, copy, modify, and distribute this            |
| software and its documentation for all purposes is hereby       |
| granted without fee, provided that the above copyright notice   |
| and this permission appear in all copies and in supporting      |
| documentation, and that the name of George Mason University     |
| not be used in advertising or publicity pertaining to           |
| distribution of the software without specific, written prior    |
| permission. GMU makes no representations about the suitability  |
| of this software for any purposes.  It is provided "AS IS"      |
| without express or implied warranties.  All risk associated     |
| with use of this software is expressly assumed by the user.     |
*----------------------------------------------------------------*/
#include "C2SIMClientLib2.h"

#pragma once

#include <iostream>
#include <boost/algorithm/string.hpp>

std::string messageType;            // CONNECTED, MESSAGE, etc
std::string messageSelector;        // Type of BML message as determined by XML matching: IBMLReport, ..
std::vector<std::string> headers;	// Raw unparsed STOMP message headers
std::map<std::string, std::string> headerMap;  // Maps header to header value 
std::string messageBody;			// body of the message not including terminating null
long messageLength;					// Length of message body not including terminating null
long contentLength;					// Length received in http header
C2SIMHeader* c2simHeader;			// Header stripped from incoming C2SIM Message

// Constructor - Initialize properties
/**
* C2SIMSTOMPMessage Constructor
*/
C2SIMSTOMPMessage::C2SIMSTOMPMessage() {
	messageSelector = "";
	messageType = "";
	headers;           // Empty vector
	headerMap;        //  intialize empty map
	messageBody = "";  // Empty message body
	messageLength = 0L;// There may not be a message body
	contentLength = 0l;// therefore no content
}


// import reference to the StompClient frame.properties
// for use as headerMap
void C2SIMSTOMPMessage::setHeaderMap(std::map<std::string, std::string> frameProperties) {
	headerMap = frameProperties;
}


// Add line to message body
void C2SIMSTOMPMessage::addToBody(std::string s) {
	messageBody += s;
}

// Getters for class properties
/**
* getMessageType - Returns the STOMP command for this message.  Normally CONNECTED or MESSAGE
* @return String - The STOMP COMMAND for this message
*/
std::string C2SIMSTOMPMessage::getMessageType() {
	return messageType;
}   // end getMessageType()

/**
* setMessageType - Saves the STOMP command for this message.  Normally CONNECTED or MESSAGE
* @param type e.g. - The STOMP COMMAND for this message
*/
void C2SIMSTOMPMessage::setMessageType(std::string type){
	messageType = type;
}// end setMessageType()

/**
* Return the BML message type determined when the server receives the message from its creator
* @return String - BML Message Selector e.g. IBML09GSR
*/
std::string C2SIMSTOMPMessage::getMessageSelector() {
	return messageSelector;
}   // end getMessageSelector

/**
* Return the body of the message, i.e. the part of the message 
* following the headers.  Does not include the terminating NULL
* @return String - The message body from the STOMP Message
*/
std::string C2SIMSTOMPMessage::getMessageBody() {
	return messageBody;
}   // end getMessageBody()

/**
* Sets value of body of the message, i.e. the part of the message following the headers.
* Does not include the terminating NULL
*/
void C2SIMSTOMPMessage::setMessageBody(std::string body){
	messageBody = body;
}// end setMessageBody()

/**
* Get the length of the message detrmined internally
* @return Long - The message length
*/
long C2SIMSTOMPMessage::getMessageLength() {
	return messageLength;
} // end getMessageLength()

/**
* set messageLength for this message
*/
void C2SIMSTOMPMessage::setMessageLength(long messageLengthParm) {
	messageLength = messageLengthParm;
}// end setMessageLength()

/**
* set contentLength for this message
*/
void C2SIMSTOMPMessage::setContentLengthFromStompHeader(long contentLengthParm) {
	contentLength = contentLengthParm;
}// end setContentLengthFromHeader()

/**
* Get the length of the message as determined by the content-length header
* @return Long - The content length
*/
 long C2SIMSTOMPMessage::getContentLength() {
	return contentLength;
}// end getContentLength()

/**
* getC2SIMHeader
* @return C2SIMHeader the c2sim header from this message
*/
 C2SIMHeader* C2SIMSTOMPMessage::getC2SIMHeader() {
	return c2simHeader;
}// end getC2SIMHeader()


 /**
 * setC2SIMHeader
 * @param C2SIMHeader the c2sim header from this essage
 */
 void C2SIMSTOMPMessage::setC2SIMHeader(C2SIMHeader* c2simHeaderParm) {
	 c2simHeader = c2simHeaderParm;
 }// end setC2SI<Header()


/**
* Get the contents of a specific STOMP header
* @param header e.g. "content-length"
* @return - String - Value of header or "" if header not set
*/
std::string C2SIMSTOMPMessage::getHeader(std::string header) {
	if (headerMap.count(header)) {
		return headerMap[header];
	}
	else {
		return "";
	}
} // end getHeader() 

// destructor
C2SIMSTOMPMessage::~C2SIMSTOMPMessage()
{
}

File: Software/Library/CS/C2SIMSDK/C2SIMClientLib/C2SIMSTOMPMessage.cs
using System.Text;

namespace C2SimClientLib;

/// <summary>
///  C2SIMSTOMPMessage
///  Encapsulates a STOMP Message along with other data created during the processing of the message
///  From original Java code implemented by Douglas Corner - George Mason University C4I and Cyber Center
public class C2SIMSTOMPMessage
{
    #region Private properties
    /// <summary>
    /// messageType  STRING - STOMP Message COMMAND received
    /// </summary>
    internal string _messageType;                   // CONNECTED, MESSAGE, etc
    /// <summary>
    /// messageSelector  STRING - Type of BML message as determined by XML matching
    /// </summary>
    internal string _messageSelector;               // IBMLReport, ..
    /// <summary>
    /// headers  Vector of STRINGs - Raw unparsed STOMP message headers
    /// </summary>
    internal List<string> _headers;
    /// <summary>
    ///  headerMap - HashMap of string,string = Maps header to header value
    /// </summary>
    internal Dictionary<string, string> _headerMap;  // Header     
    /// <summary>
    /// messageBody string = The body of the message not including terminating null
    /// </summary>
    internal string _messageBody;
    /// <summary>
    /// contentLength long - Length of message body as received not including 
    /// terminating null.  Obtained from content-length header    
    /// </summary>
    internal long _contentLength;
    /// <summary>
    /// messageLength - Length of the message after removing the C2SIM header.  
    /// For non C2SIM messages this is the same as the contentLength
    /// </summary>
    internal long _messageLength;
    /// <summary>
    /// C2SIMHeader stripped from incoming Message
    /// </summary>
    internal C2SIMHeader _c2sim;
    /// <summary>
    // Exception caught in background thread.  Used to 
    // communicate exception to background.
    // May be an otherwise empty message        
    /// </summary>
    internal Exception _error;
    #endregion

    #region Construction / teardown
    // Constructor - Initialize properties
    /// <summary>
    /// C2SIMSTOMPMessage Constructor
    /// </summary>
    public C2SIMSTOMPMessage()
    {
        _messageSelector = string.Empty;
        _messageType = string.Empty;
        _headers = new List<string>();       
        _headerMap = new Dictionary<string, string>();
        _messageBody = string.Empty;                  
        _messageLength = 0L;                
    }
    #endregion

    #region Public methods
    /// <summary>
    ///   Move the values from headers Vector creating a HashMap of header 
    ///   names and header values
    /// </summary>
    /// <returns>string - messageSelector if one was found </returns>
    public string CreateHeaderMap()
    {
        _headerMap = new Dictionary<string, string>();
        for (int i = 0; i < _headers.Count(); ++i)
        {
            int colomnIdx = _headers[i].IndexOf(':');
            if (colomnIdx < 0)
            {
                // Just ignore - some entries are not headers, e.g. 'MESSAGE'
                continue;
            }
            // Key/name before the ':'
            string header = _headers[i].Substring(0, colomnIdx);
            // Value after the ':' - may be empty
            string headerVal = _headers[i].Length > colomnIdx + 1 ? _headers[i].Substring(colomnIdx + 1) : string.Empty; 
            // Some headers may be repeated (e.g. 'destination'). Keeping the latest
            _headerMap[header] = headerVal;
            // A value of true is used with the particular message Selector for this message
            // TODO: what can be seen in some of the actual headers are entries where the header _key_ is "true",
            // for example "true:1636062688801"
            // This test for headerVal is possibly a mistake
            if (headerVal.Equals("true", StringComparison.InvariantCultureIgnoreCase))
            {
                _messageSelector = header;
            }
        }
        return _messageSelector;
    }

    /// <summary>
    /// Contents of a specific STOMP header
    /// </summary>
    /// <param name="header">Specific header e.g. "content-length"</param>
    /// <returns>- string - Value of header or string.Empty if header not set in incoming message</returns>
    public string GetHeader(string header)
    {
        return _headerMap.ContainsKey(header) ? _headerMap[header] : string.Empty;
    }
    #endregion

    #region Private methods
    /// <summary>
    /// Add a line to the list of headers
    /// </summary>
    /// <param name="s">Header to add</param>
    internal void AddHeader(string s)
    {
        _headers.Add(s);
    }

    /// <summary>
    /// Add line to message body
    /// </summary>
    /// <param name="s">Line to add</param>
    internal void AddToBody(string s)
    {
        _messageBody += s;
    }
    #endregion

    #region Public properties
    /// <summary>
    /// STOMP command for this message.  
    /// Normally CONNECTED or MESSAGE
    /// </summary>
    public string MessageType => _messageType;
    /// <summary>
    /// Message type determined when the server receives 
    /// the message from its creator 
    /// </summary>
    public string MessageSelector => _messageSelector;
    /// <summary>
    /// Body of the message, i.e. the part of the message 
    /// following the headers.  Does not include the terminating NULL
    /// </summary>
    public string MessageBody => _messageBody;
    /// <summary>
    /// Length of the message without the C2SIM Header <BR>
    ///  For non C2SIM messages this is the same as the contentLength.
    /// </summary>
    public long MessageLength => _messageLength;
    /// <summary>
    /// Length of the message as determined by the content-length header
    /// </summary>
    public long ContentLength => _contentLength;
    /// <summary>
    /// C2Sim header from this message
    /// </summary>
    public C2SIMHeader C2SIMHeader => _c2sim;

    /// <summary>
    /// Exception thrown in the stomp TCP thread, wrapped for delivery within this message
    /// </summary>
    public Exception Error => _error;
    #endregion

}
